internal/util/addmetricsannotation/addmetricsannotation.go

```diff
@@ -0,0 +1,110 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package addmetricsannotation
+
+import (
+	"fmt"
+	"os"
+	"strings"
+
+	"sigs.k8s.io/kustomize/kyaml/kio"
+	kyaml "sigs.k8s.io/kustomize/kyaml/yaml"
+)
+
+const (
+	CNRMMetricsAnnotation        = "cnrm.cloud.google.com/blueprint"
+	DisableKptMetricsEnvVariable = "DISABLE_KPT_METRICS"
+)
+
+// AddMetricsAnnotation adds metrics annotation to track the usage the of kpt
+type AddMetricsAnnotation struct {
+	// Group is the command group e.g., pkg, fn, live
+	Group string
+}
+
+// Process invokes AddMetricsAnnotation kyaml filter on the resources in input packages paths
+func Process(group string, paths ...string) error {
+	for _, path := range paths {
+		inout := &kio.LocalPackageReadWriter{PackagePath: path, PreserveSeqIndent: true, WrapBareSeqNode: true}
+		ama := &AddMetricsAnnotation{Group: group}
+		err := kio.Pipeline{
+			Inputs:  []kio.Reader{inout},
+			Filters: []kio.Filter{kio.FilterAll(ama)},
+			Outputs: []kio.Writer{inout},
+		}.Execute()
+		if err != nil {
+			// this should be a best effort, do not error if this step fails
+			// https://github.com/GoogleContainerTools/kpt/issues/2559
+			return nil
+		}
+	}
+	return nil
+}
+
+// Filter implements kyaml.Filter
+// this filter adds "cnrm.cloud.google.com/blueprint" annotation to the resource
+// if the annotation is already present, it appends kpt-<group> suffix
+// it uses "default" namespace
+func (ama *AddMetricsAnnotation) Filter(object *kyaml.RNode) (*kyaml.RNode, error) {
+	// users can opt-out by setting the "DISABLE_KPT_METRICS" environment variable
+	if os.Getenv(DisableKptMetricsEnvVariable) != "" {
```

Comment:
Shouldn't we call this in the `Process` method at the beginning? to avoid calling it for each resource ?

Better will be to wrap this in a helper method that you can invoke from both the places. something like `trackingEnabled()`

Reply:
Its better to have single entry point. So I made Filter to be non-usable by external callers. All callers should go through TrackAction() method. So no need to wrap this.